File: java/P210_1.java
class Solution {
    // dfs
    int length;
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        length = numCourses - 1;
        int[] result = new int[numCourses];
        int[] visited = new int[numCourses];
        Map<Integer, List<Integer>> map = new HashMap<Integer, List<Integer>>();
        for(int[] prerequisite: prerequisites) {
            if (!map.containsKey(prerequisite[1])) {
                map.put(prerequisite[1], new LinkedList<Integer>());
            }
            map.get(prerequisite[1]).add(prerequisite[0]);
        }
        for (int i = 0; i < numCourses; i++) {
            if (!dfs(map, visited, result, i)) {
                return new int[0];
            }            
        }
        return result;
    }
    
    private boolean dfs(Map<Integer, List<Integer>> map, int[] visited, int[] result, int index) {
        if (visited[index] == -1) {
            return false;
        }
        if (visited[index] == 1) {
            return true;
        }
        visited[index] = -1;
        if (map.containsKey(index)) {
            for (int value: map.get(index)) {
                if (!dfs(map, visited, result, value)) {
                    return false;
                }
            }
        }
        result[length--] = index;
        visited[index] = 1;
        return true;
    }
}

File: python/P210_1.py
class Solution(object):
    # bfs
    def findOrder(self, numCourses, prerequisites):
        dependence = [[] for i in range(numCourses)]
        dep_num = [0 for i in range(numCourses)]
        for each in prerequisites:
            if each[0] not in dependence[each[1]]:
                dependence[each[1]].append(each[0])
                dep_num[each[0]] += 1
        curr_list = []
        order_list = []
        for i in range(numCourses):
            if dep_num[i] == 0:
                curr_list.append(i)
        count = 0
        while curr_list:
            curr_course = curr_list[0]
            count += 1
            for i in dependence[curr_course]:
                dep_num[i] -= 1
                if dep_num[i] == 0:
                    curr_list.append(i)
            curr_list.remove(curr_course)
            order_list.append(curr_course)
        if count < numCourses:
            return []
        else:
            return order_list